components/function-controller/pkg/apis/serverless/v1alpha1/function_validation.go

```diff
@@ -45,64 +44,72 @@ type ValidationConfig struct {
 	Function     MinFunctionValues
 	BuildJob     MinBuildJobValues
 }
+type validationFunction func(*ValidationConfig) error
 
-func (fn *Function) performBasicValidation(ctx context.Context) *apis.FieldError {
-	return fn.validateObjectMeta(ctx).Also(
-		fn.Spec.validateSource(),
-		fn.Spec.validateEnv(ctx),
-		fn.Spec.validateLabels(),
-		fn.Spec.validateReplicas(ctx),
-		fn.Spec.validateFunctionResources(ctx),
-		fn.Spec.validateBuildResources(ctx),
-	)
+func (fn *Function) getBasicValidations(vc *ValidationConfig) []validationFunction {
+	return []validationFunction{
+		fn.validateObjectMeta,
+		fn.Spec.validateSource,
+		fn.Spec.validateEnv,
+		fn.Spec.validateLabels,
+		fn.Spec.validateReplicas,
+		fn.Spec.validateFunctionResources,
+		fn.Spec.validateBuildResources,
+	}
 }
 
-func (fn *Function) Validate(ctx context.Context) (errors *apis.FieldError) {
-	spec := fn.Spec
+func (fn *Function) Validate(vc *ValidationConfig) error {
+	validations := fn.getBasicValidations(vc)
 
-	if spec.Type == SourceTypeGit {
-		return fn.performBasicValidation(ctx).Also(
-			spec.validateRepository(),
-		)
+	if fn.Spec.Type == SourceTypeGit {
+		validations = append(validations, fn.Spec.validateRepository)
+	} else {
+		validations = append(validations, fn.Spec.validateDeps)
 	}
 
-	return fn.performBasicValidation(ctx).Also(
-		spec.validateDeps(),
-	)
+	return runValidations(validations, vc)
+}
+
+func runValidations(vFuns []validationFunction, vc *ValidationConfig) error {
```

Comment:
I think that `vc` should be the first parameter as it's more important that `vFuncs` and additionally the `vFuns` can be used as variadic arguments.